paasta_tools/setup_kubernetes_crd.py

```diff
@@ -0,0 +1,140 @@
+"""
+import argparse
+import logging
+import sys
+from typing import Sequence
+
+import service_configuration_lib
+from kubernetes.client.rest import ApiException
+
+from paasta_tools.kubernetes_tools import KubeClient
+from paasta_tools.kubernetes_tools import V1beta1CustomResourceDefinition
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import load_system_paasta_config
+
+log = logging.getLogger(__name__)
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description='Creates/updates kubernetes CRDs.')
+    parser.add_argument(
+        'service_list', nargs='+',
+        help="The list of services to create or update CRDs for",
+        metavar="SERVICE",
+    )
+    parser.add_argument(
+        '-c', '--cluster', dest="cluster", metavar="CLUSTER",
+        default=None,
+        help="Kubernetes cluster name",
+    )
+    parser.add_argument(
+        '-d', '--soa-dir', dest="soa_dir", metavar="SOA_DIR",
+        default=DEFAULT_SOA_DIR,
+        help="define a different soa config directory",
+    )
+    parser.add_argument(
+        '-v', '--verbose', action='store_true',
+        dest="verbose", default=False,
+    )
+    args = parser.parse_args()
+    return args
+
+
+def main() -> None:
+    args = parse_args()
+    soa_dir = args.soa_dir
+    if args.verbose:
+        logging.basicConfig(level=logging.DEBUG)
+    else:
+        logging.basicConfig(level=logging.WARNING)
+
+    if args.cluster:
+        cluster = args.cluster
+    else:
+        system_paasta_config = load_system_paasta_config()
+        cluster = system_paasta_config.get_cluster()
+
+    kube_client = KubeClient()
+
+    success = setup_kube_crd(
+        kube_client=kube_client,
+        cluster=cluster,
+        services=args.service_list,
+        soa_dir=soa_dir,
+    )
+    sys.exit(0 if success else 1)
+
+
+def setup_kube_crd(
+        kube_client: KubeClient,
+        cluster: str,
+        services: Sequence[str],
+        soa_dir: str = DEFAULT_SOA_DIR,
+) -> bool:
+    existing_crds = kube_client.apiextensions.list_custom_resource_definition()
+    success = True
+    for service in services:
+        crd_config = service_configuration_lib.read_extra_service_information(
+            service, f'kubernetes-crd-{cluster}', soa_dir=soa_dir,
+        )
+        desired_crd = V1beta1CustomResourceDefinition(
+            api_version=crd_config.get('apiVersion'),
+            kind=crd_config.get('kind'),
+            metadata=crd_config.get('metadata'),
+            spec=crd_config.get('spec'),
+            status=crd_config.get('status'),
+        )
+
+        existing_crd = None
+        for crd in existing_crds.items:
+            if crd.metadata.name == desired_crd.metadata['name']:
+                existing_crd = crd
+                break
+
+        try:
+            if existing_crd:
+                desired_crd.metadata['resourceVersion'] = existing_crd.metadata.resource_version
```

Comment:
though this might be the right field for us to use, I'm not sure I necessarily want users to have to specify the resource version for the CRD. Can we instead take a hash of the CRD to determine the resource version, and use that?

Reply:
as i understand resourceVersion is opaque value maintained by the server. it's used for concurrency control and clients aren't supposed to modify it. https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#concurrency-control-and-consistency